src/PerformanceSensitiveAnalyzers/Core/ExplicitAllocationAnalyzer.cs

```diff
@@ -0,0 +1,112 @@
+        private static readonly LocalizableString s_localizableLetCauseRuleTitleAndMessage = new LocalizableResourceString(nameof(PerformanceSensitiveAnalyzersResources.LetCauseRuleTitleAndMessage), PerformanceSensitiveAnalyzersResources.ResourceManager, typeof(PerformanceSensitiveAnalyzersResources));
+
+        internal static DiagnosticDescriptor ArrayCreationRule = new DiagnosticDescriptor(
+            ArrayCreationRuleId,
+            s_localizableArrayCreationRuleTitleAndMessage,
+            s_localizableArrayCreationRuleTitleAndMessage,
+            DiagnosticCategory.Performance,
+            DiagnosticSeverity.Info,
+            isEnabledByDefault: true);
+
+        internal static DiagnosticDescriptor ObjectCreationRule = new DiagnosticDescriptor(
+            ObjectCreationRuleId,
+            s_localizableObjectCreationRuleTitleAndMessage,
+            s_localizableObjectCreationRuleTitleAndMessage,
+            DiagnosticCategory.Performance,
+            DiagnosticSeverity.Info,
+            isEnabledByDefault: true);
+
+        internal static DiagnosticDescriptor AnonymousObjectCreationRule = new DiagnosticDescriptor(
+            AnonymousObjectCreationRuleId,
+            s_localizablAnonymousObjectCreationRuleTitleAndMessage,
+            s_localizablAnonymousObjectCreationRuleTitleAndMessage,
+            DiagnosticCategory.Performance,
+            DiagnosticSeverity.Info,
+            isEnabledByDefault: true,
+            helpLinkUri: "http://msdn.microsoft.com/en-us/library/bb397696.aspx");
+
+        internal static DiagnosticDescriptor LetCauseRule = new DiagnosticDescriptor(
+            LetCauseRuleId,
+            s_localizableLetCauseRuleTitleAndMessage,
+            s_localizableLetCauseRuleTitleAndMessage,
+            DiagnosticCategory.Performance,
+            DiagnosticSeverity.Info,
+            isEnabledByDefault: true);
+
+        private static readonly object[] EmptyMessageArgs = Array.Empty<object>();
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+            ArrayCreationRule,
+            ObjectCreationRule,
+            AnonymousObjectCreationRule,
+            LetCauseRule);
+
+        protected override ImmutableArray<OperationKind> Operations => ImmutableArray.Create(
+            OperationKind.ArrayCreation,
+            OperationKind.ObjectCreation,
+            OperationKind.AnonymousObjectCreation);
+
+        protected override void AnalyzeNode(OperationAnalysisContext context, in PerformanceSensitiveInfo info)
+        {
+            if (context.Operation is IArrayCreationOperation)
+            {
+                context.ReportDiagnostic(Diagnostic.Create(ArrayCreationRule, context.Operation.Syntax.GetLocation(), EmptyMessageArgs));
```

Comment:
❗️ This should not be reported when `IArrayCreationOperation.IsImplicit` is true. Some items excluded by this rule will already be reported as `params` array allocations. However, I'm still trying to determine if this would result in cases where array allocations are omitted from analysis.